pkg/indexers/indexers.go

```diff
@@ -142,3 +159,21 @@ func ByPathAndName[T runtime.Object](groupResource schema.GroupResource, indexer
 	}
 	return objs[0].(T), nil
 }
+
+// ByPathAndNameWithFallback returns the instance of T from the indexer with the matching path and name. Path may be a canonical path
+// or a cluster name. Note: this depends on the presence of the optional "kcp.io/path" annotation. If no instance is found, globalIndexer
+// is searched as well. Any errors short-circuit this logic.
+func ByPathAndNameWithFallback[T runtime.Object](groupResource schema.GroupResource, indexer, globalIndexer cache.Indexer, path logicalcluster.Path, name string) (ret T, err error) {
```

Comment:
Why do we need both `ByIndexWithFallback` and `ByPathAndNameWithFallback` methods? 
It seems that both methods eventually make use of `indexer.ByIndex` :)

The difference seems to be that `ByIndex` returns an array of object whereas `ByPathAndName` returns a single object.

Do we need this distinction or could we unify into one method?

Reply:
@p0lyn0mial please feel free to propose a simpler factoring! One was not obvious to me since the error checking is different.